torch/csrc/jit/script/compiler.cpp

```diff
@@ -1035,50 +1041,27 @@ struct to_ir {
     }
     return arguments;
   }
-  std::vector<Argument> emitReturn(c10::optional<Return> return_stmt_, const FunctionSchema& schema) {
+
+  Argument emitReturn(const SourceRange& range, c10::optional<Return> return_stmt, const FunctionSchema& schema) {
+    JIT_ASSERT(schema.returns().size() <= 1);
     // outputs
-    std::vector<Argument> returns;
-    if (return_stmt_) {
-      auto return_stmt = *return_stmt_;
-      auto results = getValues(return_stmt.values(), true);
-      // a single return value that is a tuple expands in place:
-      // return a
-      if (return_stmt.values().size() == 1 && results.size() == 1) {
-        auto result = results.at(0);
-        if(result->type()->cast<TupleType>()) {
-          results = createTupleUnpack(result).vec();
-        }
-      }
-      if (!schema.is_varret() && schema.returns().size() != results.size()) {
-        throw ErrorReport(def.range()) << "Number of type annotations for function"
-          << " return (" << schema.returns().size() << ") does not match"
-          << " the number of returns from the function (" << results.size() << ")!";
-      }
-      const auto& range = return_stmt.range();
-      size_t return_type_idx = 0;
-      for (auto r : results) {
-        TypePtr type = DynamicType::get();
-        if (!schema.is_varret()) {
-          type = schema.returns().at(return_type_idx).type();
-          r = tryConvertToType(range, *graph, type, r, /*allow_conversions=*/false);
-          if (!r->type()->isSubtypeOf(type)) {
-            throw ErrorReport(return_stmt.range()) << "Return value at position "
-              << return_type_idx << " was annotated as having type " << type->str()
-              << " but is actually of type " << r->type()->str();
-          }
-          return_type_idx++;
-        }
-        graph->registerOutput(r);
-        returns.emplace_back("", type);
-      }
-    } else if (schema.returns().size() > 0) {
-      // schema has returns but there's no return nodes in graph
-      throw ErrorReport() << "Expected " << schema.returns().size()
-                          << " return value"
-                          << (schema.returns().size() > 1 ? "s" : "")
-                          << " but found no return statement";
+    Value* result = return_stmt ? emitExpr(return_stmt->expr())
+                                : graph->insertConstant(IValue(), range);
+    TypePtr result_type = schema.returns().size() > 0
```

Comment:
Confused by this logic. If a schema has multiple returns, why is `result_type` set to the type of the first one?

Reply:
I think schema can only have one return or None now as stated by the `JIT_ASSERT` stmt.

Reply:
```
    // we represent no annoation on a return type as having no values in the
     // schema's return() list
     // in emitReturn we take the actual return value to be the value of the return
     // statement if no one was provided here
```
Ill add a back reference.

Reply:
In other words, we need an encoding for "user has not annotated the return type, infer it", and using schema.returns() being empty (versus having exactly 1 element) seemed to be the simplest thing. Previously this was encoded using `is_varret` but the implementation was buggy and it allowed incorrect types to be returned.